.github/docs/CONFIGURATION.md
# Configuration

Configuring the resource is quite simple. 

All of the options are in the `shared/config.lua` file.


## Global Config Options
| Field Name       	| Type      	| Default                    	| Description                                                                                                                                	|
|------------------	|-----------	|----------------------------	|--------------------------------------------------------------------------------------------------------------------------------------------	|
| DebugMode        	| `boolean` 	| `false`                    	| When enabled additional console logging done and development UI is shown.                                                                  	|
| GroundRemoteRope 	| `boolean` 	| `false`                    	| Forces the rope for the remote to not clip through the ground, this can reduce performance and make the script run a lot more expensively. 	|
| MaxHookReach     	| `number`  	| `10.0`                     	| The max distance the hook will be able to be attached from.                                                                                	|
| RemoteModel      	| `hash`    	| `w_am_digiscanner_reh` 	    | The hash for the remote prop                                                                                                               	|
| HookModel        	| `hash`    	| `prop_v_hook_s`        	    | The hash for the hook prop                                                                                                                 	|
| TowTrucks        	| `object`   	| N/A                        	| The config for each of the supported tow trucks                                                                                            	|


## Tow Trucks Config Options

There are 2 different types of tow trucks supported by the script, **scoop based** and **prop based**, each of which have their own unique set of config options.

To select the type of tow truck for a particular entry, use the field `truckType` with either the string `"scoop"` or `"prop"` for their respective types.


### Scoop Based Tow Trucks
Scoop based tow trucks are tow trucks where the bed can be controlled with the same keybinds as the scoop on a dozzer.

*As of version 1.0.0, scoop based tow trucks are not yet supported!*
| Field Name       	| Type                                                      	| Default                                                                     	| Description                                                                                                                                 	|
|------------------	|-----------------------------------------------------------	|-----------------------------------------------------------------------------	|---------------------------------------------------------------------------------------------------------------------------------------------	|
| truckType        	| `string`                                                  	| N/A                                                                         	| The mode of the tow truck. Always needs to be `"scoop"` for scoop based tow trucks.                                                         	|
| truckModel       	| `hash`                                                    	| N/A                                                                         	| The model hash for the tow truck vehicle.                                                                                                   	|
| controlBoxOffset 	| `vector3`                                                 	| `vector3(-1.05, -1.0, 0.0)`                                                 	| The offset from the vehicle where the control box for the bed is located. It's this location where the truck can be controlled by the user. 	|
| hookRoot         	| {<br>  boneName = `string`,<br>  offset = `vector3`,<br>} 	| {<br>  boneName = "attach_male",<br>  offset = vector3(0.0, 0.0, 0.0),<br>} 	| The offset from the a bone on the truck, where the rope for the hook starts.                                                                	|
| bedAttach        	| {<br>  boneName = `string`,<br>  offset = `vector3`,<br>} 	| {<br>  boneName = "misc_z",<br>  offset = vector3(0.0, 0.0, 0.70),<br>}     	| The offset from a bone on the truck where the cars will be attached to.                                                                     	|
| bedPositions     	| {<br>  raised = `number`,<br>  lowered = `number`,<br>}   	| N/A                                                                         	| The scoop arm value for where the bed is lowered all the way and raised all the way.                                                        	|


### Prop Based Tow Trucks
Prop based tow trucks are tow trucks where the bed is togglable via a vehicle extra and where the bed is also streamed as a spawnable object.

| Field Name          	| Type                                                                                                                                                                                                                          	| Default                                                                                                                                                                                                                                                                                                           	| Description                                                                                                                                                                                                                                                                                                                                                            	|
|---------------------	|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------	|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------	|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------	|
| truckType           	| `string`                                                                                                                                                                                                                       	| N/A                                                                                                                                                                                                                                                                                                               	| The mode of the tow truck. Always needs to be `"prop"` for prop based tow trucks.                                                                                                                                                                                                                                                                                      	|
| truckModel          	| `hash`                                                                                                                                                                                                                        	| N/A                                                                                                                                                                                                                                                                                                               	| The model hash for the tow truck vehicle.                                                                                                                                                                                                                                                                                                                              	|
| bedModel            	| `hash`                                                                                                                                                                                                                        	| N/A                                                                                                                                                                                                                                                                                                               	| The model hash of the bed prop                                                                                                                                                                                                                                                                                                                                         	|
| bedExtraIndex       	| `number`                                                                                                                                                                                                                      	| N/A                                                                                                                                                                                                                                                                                                               	| The extra number for toggling the bed on the vehicle                                                                                                                                                                                                                                                                                                                   	|
| lerpMult            	| `float`                                                                                                                                                                                                                       	| `4.0`                                                                                                                                                                                                                                                                                                             	| The multiplier applied to the lerp calculation. In other words, this is a speed multiplier to the raise and lower process.                                                                                                                                                                                                                                             	|
| hasRemote           	| `boolean`                                                                                                                                                                                                                     	| `true`                                                                                                                                                                                                                                                                                                            	| If the truck has a remote that can be grabbed to control the bed movement                                                                                                                                                                                                                                                                                              	|
| remoteStorageOffset 	| `vector3`                                                                                                                                                                                                                     	| `vector3(-1.05, -1.0, 0.0)`                                                                                                                                                                                                                                                                                       	| The offset from the vehicle where the remote can be retrieved from.<br>*Only used if "hasRemote" is set to true.*                                                                                                                                                                                                                                                      	|
| controlBoxOffset    	| `vector3`                                                                                                                                                                                                                     	| `vector3(-1.05, -1.0, 0.0)`                                                                                                                                                                                                                                                                                       	| The offset from the vehicle where the control box for the bed is located. It's this location where the truck can be controlled by the user.<br>*Only used if "hasRemote" is set to false.*                                                                                                                                                                             	|
| hookRootOffset      	| `vector3`                                                                                                                                                                                                                     	| `vector3(0.025, 4.5, 0.1)`                                                                                                                                                                                                                                                                                        	| The offset from the bed on where the rope for the hook starts.                                                                                                                                                                                                                                                                                                         	|
| bedAttachOffset     	| `vector3`                                                                                                                                                                                                                     	| `vector3(0.0, 1.5, 0.3)`                                                                                                                                                                                                                                                                                          	| The offset from the bed that would attach a car centered onto the bed.                                                                                                                                                                                                                                                                                                 	|
| bedOffsets          	| {<br>  raised = {<br>    pos = `vector3`,<br>    rot = `vector3`,<br>  },<br>  back = {<br>    pos = `vector3`,<br>    rot = `vector3`,<br>  },<br>  lowered = {<br>    pos = `vector3`,<br>    rot = `vector3`,<br>  },<br>} 	| {<br>  raised = {<br>    pos = vector3(0.0, -3.8, 0.45),<br>    rot = vector3(0.0, 0.0, 0.0),<br>  },<br>  back = {<br>    pos = vector3(0.0, -4.0, 0.0),<br>    rot = vector3(0.0, 0.0, 0.0),<br>  },<br>  lowered = {<br>    pos = vector3(0.0, -0.4, -1.0),<br>    rot = vector3(12.0, 0.0, 0.0),<br>  },<br>} 	| The offsets which control how the bed slides from raised to lowered.<br>When a bed is in default position it's using the "raised" offset. <br>In the process of lowering the bed the bed will start at the "raised" offset, <br>slide back to the "back" offset and then tilt to the "lowered" offset. <br>This process is reversed when going from lowered to raised. 	|


### Some Pre-Made Configs


---
#### Biff Flatbed 2x Boxe by MFD
Tebex Store Page: https://mfd.tebex.io/package/6281210

Config:
```lua
[`biftowmfd2`] = {
    truckType = "scoop",
    truckModel = `biftowmfd2`,
    controlBoxOffset = vector3(-1.1, -1.3, -0.4),
    hookRoot = {
        boneName = "attach_male",
        offset = vector3(0.0, 0.0, 0.0),
    },
    bedAttach = {
        boneName = "misc_z",
        offset = vector3(0.0, 0.0, 0.70),
    },
    bedPositions = {
        raised = 0.0,
        lowered = 0.25,
    },
},
```

DevJacob_Towing/shared/config.lua
Config = {}

Config["DebugMode"] = false

--[[
	Forces the rope for the remote to not clip through the ground,
	this can reduce performance and make the script run a lot more expensively
]]
Config["GroundRemoteRope"] = false


Config["MaxHookReach"] = 10.0


Config["RemoteModel"] = `w_am_digiscanner_reh`


Config["HookModel"] = `prop_v_hook_s`


Config["TowTrucks"] = {
	[`flatbed3`] = {
		truckType = "prop",
		truckModel = `flatbed3`,
		bedModel = `flatbed3_base`,
		bedExtraIndex = 1,
		lerpMult = 4.0,
		hasRemote = true,
		remoteStorageOffset = vector3(-1.05, -1.0, 0.0),
		controlBoxOffset = vector3(-1.05, -1.0, 0.0),
		hookRootOffset = vector3(0.025, 4.5, 0.1),
		bedAttachOffset = vector3(0.0, 1.5, 0.3),
		bedOffsets = {
			raised = {
				pos = vector3(0.0, -3.8, 0.45),
				rot = vector3(0.0, 0.0, 0.0),
			},
			back = {
				pos = vector3(0.0, -4.0, 0.0),
				rot = vector3(0.0, 0.0, 0.0),
			},
			lowered = {
				pos = vector3(0.0, -0.4, -1.0),
				rot = vector3(12.0, 0.0, 0.0),
			},
		},
	},

	[`biftowmfd2`] = {
		truckType = "scoop",
		truckModel = `biftowmfd2`,
		controlBoxOffset = vector3(-1.1, -1.3, -0.4),
		hookRoot = {
			boneName = "attach_male",
			offset = vector3(0.0, 0.0, 0.0),
		},
		bedAttach = {
			boneName = "misc_z",
			offset = vector3(0.0, 0.0, 0.70),
		},
		bedPositions = {
			raised = 0.0,
			lowered = 0.25,
		},
	},
}